File: leetcode/735/LeetCode 735: AsteroidsCollision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        remaining_asteroids = []
        
        for asteroid in asteroids:
            if asteroid > 0:
                remaining_asteroids.append(asteroid)
                continue            
            
            while remaining_asteroids and 0 < remaining_asteroids[-1] < abs(asteroid):
                remaining_asteroids.pop()
                
            if not remaining_asteroids or remaining_asteroids[-1] < 0:
                remaining_asteroids.append(asteroid)
                continue
            
            if remaining_asteroids[-1] == abs(asteroid):
                remaining_asteroids.pop()
                continue               
            
                
        
        return remaining_asteroids

File: leetcode/LeetCode 735: AsteroidsCollision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> remainingAsteroids;
        
        for(auto asteroid : asteroids){
            if(asteroid > 0){
                remainingAsteroids.push(asteroid);
                continue;
            }
            
            while(!remainingAsteroids.empty() && remainingAsteroids.top() > 0 && remainingAsteroids.top() < abs(asteroid) ){
                remainingAsteroids.pop();                
            }
            
            if (remainingAsteroids.empty() || remainingAsteroids.top() < 0 ){
                remainingAsteroids.push(asteroid);
                continue;
            }
            
            if ( remainingAsteroids.top() == abs(asteroid) ){
                remainingAsteroids.pop();
                continue;
            }
            
            
        }
        
        
        vector<int> asteroidsResult;
        while(!remainingAsteroids.empty()){
            asteroidsResult.push_back(remainingAsteroids.top());
            remainingAsteroids.pop();
        }
        
        reverse(asteroidsResult.begin(), asteroidsResult.end());
        
        return asteroidsResult;
    }
};
